compiler/rustc_trait_selection/src/traits/project.rs

```diff
@@ -412,37 +427,40 @@ impl<'a, 'b, 'tcx> TypeFolder<'tcx> for AssocTypeNormalizer<'a, 'b, 'tcx> {
                 // give up and fall back to pretending like we never tried!
 
                 let infcx = self.selcx.infcx();
-                let (data, mapped_regions, mapped_types, mapped_consts, universe_map) =
-                    BoundVarReplacer::replace_bound_vars(infcx, data);
-
-                let normalized_ty = opt_normalize_projection_type(
-                    self.selcx,
-                    self.param_env,
-                    data,
-                    self.cause.clone(),
-                    self.depth,
-                    &mut self.obligations,
-                )
-                .ok()
-                .flatten()
-                .unwrap_or_else(|| ty);
-
-                let normalized_ty = PlaceholderReplacer::replace_placeholders(
-                    infcx,
-                    mapped_regions,
-                    mapped_types,
-                    mapped_consts,
-                    universe_map,
-                    normalized_ty,
-                );
-                debug!(
-                    ?self.depth,
-                    ?ty,
-                    ?normalized_ty,
-                    obligations.len = ?self.obligations.len(),
-                    "AssocTypeNormalizer: normalized type"
-                );
-                normalized_ty
+                let replaced =
+                    BoundVarReplacer::replace_bound_vars(infcx, &mut self.universes, data);
+                if let Some((data, mapped_regions, mapped_types, mapped_consts)) = replaced {
+                    let normalized_ty = opt_normalize_projection_type(
```

Comment:
I think we should use a call to [`commit_if_ok`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_infer/infer/struct.InferCtxt.html#method.commit_if_ok) here, so that we unroll the inference effects in the event of error or ambiguity. We'd have to translate `Option` to `Error` though